### ISSUE TITLE
Error when using PMEM_AND_DISK storage level

### ISSUE THREAD

**Issue by rophy1310** [oap-project/pmem-spill#56]
Hello. I'm trying to cache RDD in my persistent memory and I continuously get this error. 

![image](https://user-images.githubusercontent.com/39511546/127774217-f0f24bff-4db0-460e-9019-261c3c84d317.png)

My system is a single NUMA node system and I have one pmem path. 
I installed spark version 3.0.0 and hadoop version 3.2.0 so that I can use pmem-spill and pmem-common v1.1.0-spark-3.0.0.

Because of the error message as above, I first thought that the reason of the problem was because of the memkind library. However I tested the memkind library with the test examples provided and checked that memkind_malloc worked fine.

Is there something I'm missing? Thank you.

**Comment by yeyuqiang** [comment]
@rophy1310 Are you trying the release jar file? If so, please try to build from source code following the doc and check whether libmemkind.so is under LD_LIBRARY_PATH.

**Comment by rophy1310** [comment]
I followed the developer guide, built the OAP from the source code, and also checked libmemkind.so is under LD_LIBRARY_PATH, but still had the same problem. So I checked the dependency installation script(prepare_oap_env.sh) provided at oap-tools and find out that it is better to use the oap-tools master branch for Ubuntu. After using this version, the problem seems to be solved. So now I am using Spark ver 3.1.1, and Hadoop ver 3.2.0.

But now I'm having another problem: java.lang.NoClassDefFoundError: com/intel/oap/common/unsafe/PersistentMemoryPlatform

I checked the pmem-common/target/classes/com/intel/oap/common/unsafe directory and there was PersistentMemoryPlatform class in there, so I'm not sure why this error occurs.

**Comment by yeyuqiang** [comment]
Did you add pmem-commom to spark.executor.extraClassPath? See the detail: https://github.com/oap-project/pmem-spill#configuration

**Comment by rophy1310** [comment]
Yes. I added pmem-common to spark.executor.extraClassPath as the below screenshot. I added the configurations in spark-defaults.conf but java.lang.NoClassDefFoundError: com/intel/oap/common/unsafe/PersistentMemoryPlatform occurs.

![image](https://user-images.githubusercontent.com/39511546/129315600-b9f37716-46f7-478d-922b-cccd47247582.png)

**Comment by yeyuqiang** [comment]
How you run spark command and how you run your workload with PMem? It's better to run spark on yarn with these configuration.

**Comment by rophy1310** [comment]
I was using standalone mode in spark. The spark command I'm using is the K-means benchmark which is mentioned in the https://oap-project.github.io/pmem-spill/1.1.1/ . 

The above problem was solved after changing to yarn mode and running the command as the root user. Thank you.

But now I'm having an error: java.lang.ExceptionInInitializerError: Persistent memory initialize (path: /mnt/pmem0/executor_1:, size: 248) failed. Please check the path permission and initial size. (Please refer to the screenshot below.)
![image](https://user-images.githubusercontent.com/39511546/130452342-fd74b2a9-cfc9-4a78-a5be-e8d94d740f78.png)


I checked the initialize size and path, but it looked okay to me because the path is /mnt/pmem0 and the size is 256GB. At first, I thought that the size was the problem so I tried it with both values 248(which was the size I could check through command lsblk), 256 but still the problem occurs. 

I run the command as root user but could the problem occur because of the path permission?

**Comment by rophy1310** [comment]
I fixed the above problem by changing the configuration in spark-defaults.conf. 

It is said to put my Optane PMem size in GB so I put the value 256 as below.

...
spark.memory.pmem.initial.path /mnt/pmem0
spark.memory.pmem.initial.size 256
...

But somehow, it doesn't converts to Bytes through conf.getSizeAsBytes("spark.memory.pmem.initial.size",0L) in Blockmanager.scala.

So I converted the value to Bytes as below.

...
spark.memory.pmem.initial.path /mnt/pmem0
spark.memory.pmem.initial.size 274877906944
...

And the problem was solved. Now the K-means benchmark works. Thank you very much.
